# Введение в искусственные нейронные сети
# Урок 3. Keras

## Практическое задание

<ol>
    <li>Попробуйте обучить нейронную сеть на Keras с другими параметрами. 
        Опишите в комментарии к уроку - какой результата вы добились от нейросети? Что помогло вам улучшить ее точность?</li>
</ol>

In [36]:
# The full neural network code!
###############################
import numpy as np
import mnist
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from keras.optimizers import SGD

train_images = mnist.train_images()
train_labels = mnist.train_labels()
test_images = mnist.test_images()
test_labels = mnist.test_labels()

# Normalize the images.
train_images = (train_images / 255) - 0.5
test_images = (test_images / 255) - 0.5

# Flatten the images.
train_images = train_images.reshape((-1, 784))
test_images = test_images.reshape((-1, 784))

# Build the model.
model = Sequential([
  Dense(64, activation='relu', input_shape=(784,)),
  Dense(64, activation='relu'),
  Dense(10, activation='softmax'),
])

sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
# Compile the model.
model.compile(
  optimizer=sgd,
  loss='categorical_crossentropy',
  metrics=['accuracy'],
)

# Train the model.
model.fit(
  train_images,
  to_categorical(train_labels),
  epochs=5,
  batch_size=34,
)

# Evaluate the model.
model.evaluate(
  test_images,
  to_categorical(test_labels)
)

# Save the model to disk.
model.save_weights('model.h5')

# Load the model from disk later using:
# model.load_weights('model.h5')

# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

# Print our model's predictions.
print(np.argmax(predictions, axis=1)) # [7, 2, 1, 0, 4]

# Check our predictions against the ground truths.
print(test_labels[:5]) # [7, 2, 1, 0, 4]

Epoch 1/5
60000/60000 [==============================] - 3s 51us/step - loss: 0.3326 - accuracy: 0.8989
Epoch 2/5
60000/60000 [==============================] - 3s 50us/step - loss: 0.1670 - accuracy: 0.9488
Epoch 3/5
60000/60000 [==============================] - 3s 50us/step - loss: 0.1279 - accuracy: 0.9605
Epoch 4/5
60000/60000 [==============================] - 3s 51us/step - loss: 0.1078 - accuracy: 0.9670
Epoch 5/5
10000/10000 [==============================] - 0s 26us/step
[7 2 1 0 4]
[7 2 1 0 4]


1. Попробовал раличные функции активации tanh, sigmoid, hard_sigmoid. Качество практически не изменилось.  
С функцией активации exponential accuracy стало 0.0987.
2. Количество нейронов в слое влияет на качество.
3. Изменение параметра optimizer на SGD, время обучения сокращается на 40 секунд, а качество не изменяется.  
optimizer на Adamax время обучения меньше, но после завершения 5 эпохи, качество становится:  
accuracy: 0.96
4. Количество эпох вляет на точность предсказания модели
5. batch_size также влияет на точность нейронной сети

#### 2. Поработайте с документацией Keras.

1. get_layer(name=None, index=None)
Используется для того, чтобы извлечь необходмый слой из нейронной сети. Это можно применить, например, если нам необходимо посмотреть выход с какого-нибудь внутреннего слоя. Или мы хотим переместь какой-то слой из одной модели в другую.
2. У Keras есть Text Preprocessing(Предварительная обработка текста)  
Слова называются токенами, а процесс разбиения текста на токены называется токенизацией. 
Методы Keras для подготовки текстовых данных для глубокого обучения:  
  __Tokenizer__  
  Позволяет векторизовать текс, превращая каждый текст в последовательность целых чисел (каждое целое число является индексом токена в словаре) или в вектор, в котором коэффициент для каждого токена может быть двоичным на основе количества слов.  
 После подбора токенизатор предоставляет 4 атрибута, которые вы можете использовать для запроса того, что вы узнали о ваших документах:
    word_counts: Словарь слов и их количество.  
    word_docs: Словарь слов и сколько документов каждый появился в.  
    word_index: Словарь слов и их уникально назначенных целых чисел.  
    DOCUMENT_COUNT: Целое число от общего числа документов, которые были использованы для размещения токенизатора.  
      
  __hashing_trick__  
  Преобразование односторонней хеш-функцией слов в целые числа.

  __one_hot__  
  функция является оберткой для функции hashing_trick (). Функция возвращает целочисленную версию документа. Использование хеш-функции означает, что возможны коллизии, и не всем словам будут присвоены уникальные целочисленные значения.
  
  __text_to_word_sequence__
разбивает текст на список слов.  
По умолчанию эта функция автоматически делает 3 вещи:
   1. Разбивает слова по пробелам (split = »«).  
   2. Отфильтровывает пунктуацию (filters = ’!» # $% & Amp; () * +, -. /:; & Lt; = & gt;? @ [\\] ^ _ `{|} ~ \ t \ n’).  
   3. Преобразует текст в нижний регистр (нижний = True).  
